python/Ganga/new_tests/GPI/TestSubjobs.py

```diff
@@ -0,0 +1,59 @@
+from GangaUnitTest import GangaUnitTest
+
+class TestSubjobs(GangaUnitTest):
+
+    def setUp(self):
+        """Make sure that the Job object isn't destroyed between tests"""
+        super(TestSubjobs, self).setUp()
+        from Ganga.Utility.Config import setConfigOption
+        setConfigOption('TestingFramework', 'AutoCleanup', 'False')
+
+    def testLargeJobSubmission(self):
+        """
+        Create lots of subjobs and submit it
+        """
+        from Ganga.GPI import Job, GenericSplitter, Local
+        j = Job()
+        j.application.exe = "sleep"
+        j.splitter = GenericSplitter()
+        j.splitter.attribute = 'application.args'
+        j.splitter.values = [ ['400'] for i in range(0, 20) ]
+        j.backend = Local()
+        j.submit()
+
+        self.assertEqual( len(j.subjobs), 20 )
+
+    def testSetParentOnLoad(self):
+        """
+        Test that the parents are set correctly on load
+        """
+        from Ganga.GPI import jobs, queues, Executable, Local
+        from Ganga.GPIDev.Base.Proxy import isType
+        import random
+        import string
+
+        def flush_full_job():
+            j = jobs(0)
+            j.comment = "Make sure I'm dirty " + ''.join( random.choice( string.ascii_uppercase) for _ in range(5))
+            j._impl._getRegistry()._flush([j])
+
+        # Make sure the main job is fully loaded
+        j = jobs(0)
+        self.assertTrue( isType(j.application, Executable) )
+        self.assertTrue( isType(j.backend, Local) )
+        self.assertEqual( j.application.exe, "sleep" )
+
+        # fire off a load of threads to flush
+        for i in range(0, 20):
+            queues.add( flush_full_job )
+
+        # Now loop over and force the load of all the subjobs
+        for sj in j.subjobs:
+            self.assertEqual( sj.splitter, None )
+            self.assertTrue( isType(sj.application, Executable) )
+            self.assertTrue( isType(sj.backend, Local) )
+            self.assertEqual( sj.application.exe, "sleep" )
+            self.assertEqual( sj.application.args, ['400'] )
+            self.assertIs( sj._impl._getRoot(), j._impl)
```

Comment:
re. `self.assert...()`: when we first started porting over tests we were using `unittest` and `nose` to provide feedback. They use a style ported from `JUnit` (i.e. Java) for their asserts. They provide a lot more information than plain `assert`s in that situation and so I was encouraging their usage. Now that we are using pytest it has a system to interpret `assert` statements and extract the information in a much cleverer way.

The upshot is that before we *had* to use `self.assert...()` to get good messages but now we're using pytest we don't *have* to any more. I'm not very fussed either way but if we can get as good feedback from a simpler and cleaner `assert foo is bar` then that *slightly* wins out in my mind. There's no imperative to change anything but sometimes changing them if you're doing other changes anyway might be worth it.